sled-agent/src/bootstrap/early_networking.rs

```diff
@@ -38,59 +47,238 @@ pub enum EarlyNetworkSetupError {
     Dendrite(String),
 
     #[error("Error during DNS lookup: {0}")]
-    DnsResolver(#[from] internal_dns::resolver::ResolveError),
+    DnsResolver(#[from] ResolveError),
+}
+
+enum LookupSwitchZoneAddrsResult {
+    // We found every switch zone reported by internal DNS.
+    TotalSuccess(HashMap<SwitchLocation, Ipv6Addr>),
+    // We found some (but not all) switch zones reported by internal DNS.
+    PartialSuccess(HashMap<SwitchLocation, Ipv6Addr>),
 }
 
 /// Code for configuring the necessary network bits to bring up the control
 /// plane
-pub struct EarlyNetworkSetup {
-    log: Logger,
+pub struct EarlyNetworkSetup<'a> {
+    log: &'a Logger,
 }
 
-impl EarlyNetworkSetup {
-    pub fn new(log: &Logger) -> Self {
-        EarlyNetworkSetup { log: log.clone() }
+impl<'a> EarlyNetworkSetup<'a> {
+    pub fn new(log: &'a Logger) -> Self {
+        EarlyNetworkSetup { log }
+    }
+
+    pub async fn lookup_boundary_switch_addrs(
+        &self,
+        resolver: &DnsResolver,
+        config: &RackNetworkConfig,
+    ) -> HashSet<Ipv6Addr> {
+        let switch_zone_addrs = self.lookup_switch_zone_addrs(resolver).await;
+        let mut boundary_switch_addrs = HashSet::new();
```

Comment:
This may be just me, but I found myself confused by the name `boundary_switch_addrs`. I had to read the code to figure out that this meant: "the underlay address of a a switch zone for a switch which has an uplink".  I thought a better name would imply all those things, even though it might be a bit long. My best suggestion was `switch_zone_underlay_addr_with_uplink`, even though `switch_zone_underlay_addr_for_switch_with_uplink` is even clearer, but also insanely verbose.

After thinking about this a bit more, I think a block comment at the top of `EarlyNetworkSetup` that explains what it is intended for and defines terms like `boundary_switch` and `uplink` would be most appropriate. Then we can just keep `boundary_switch_addrs` as is. I'd be happy to help write that block comment tomorrow if desired as it's getting late here.
Further thoughts:
Does `boundary_switch` always mean a switch with an uplink? In the common, and preferred case, won't both switches be `boundary_switches` by these definitions?

Reply:
> In the common, and preferred case, won't both switches be boundary_switches by these definitions?
 
In a single rack setup, yes. In a multitrack setup, there may be a plethora of switches just functioning as internal transit between racks.

Reply:
> > In the common, and preferred case, won't both switches be boundary_switches by these definitions?
> 
> In a single rack setup, yes. In a multitrack setup, there may be a plethora of switches just functioning as internal transit between racks.

Aha. Thanks. Glad I asked.